tests/lib/hardware/libvirt.py

```diff
@@ -0,0 +1,300 @@
+<address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
+</interface>
+<serial type='pty'>
+<target port='0'/>
+</serial>
+<console type='pty'>
+<target type='serial' port='0'/>
+</console>
+<channel type='spicevmc'>
+<target type='virtio' name='com.redhat.spice.0'/>
+<address type='virtio-serial' controller='0' bus='0' port='1'/>
+</channel>
+<input type='mouse' bus='ps2'/>
+<input type='keyboard' bus='ps2'/>
+<graphics type='spice' autoport='yes'/>
+<video>
+<model type='vga'/>
+</video>
+<redirdev bus='usb' type='spicevmc'>
+</redirdev>
+<memballoon model='virtio'>
+<address type='pci' domain='0x0000' bus='0x00' slot='0x06' function='0x0'/>
+</memballoon>
+</devices>
+</domain>
+        """ % {
+            "domain_name": domain_name, "image": image,
+            "cloud_init_seed": cloud_init_seed,
+            "network_name": network_name,
+            "memory": memory
+        }
+
+
+class Hardware(HardwareBase):
+    def __init__(self):
+        super().__init__()
+        self._network = self.conn.networkLookupByName(
+            config.PROVIDER_LIBVIRT_NETWORK)
+        if not self._network:
+            raise Exception(f'Can not get libvirt network '
+                            '{config.PROVIDER_LIBVIRT_NETWORK}')
+        logger.info(f"Got libvirt network {self._network.name()}")
+
+    def get_connection(self):
+        conn = libvirt.open(config.PROVIDER_LIBVIRT_CONNECTION)
+        if not conn:
+            raise Exception(f'Can not open libvirt connection '
+                            '{config.PROVIDER_LIBVIRT_CONNECTION}')
+        logger.debug(f"Got connection to libvirt: {conn}")
+        return conn
+
+    def _boot_node(self, name: str, role: NodeRole, tags: List[str]):
+        # get a fresh connection to avoid threading problems
+        conn = self.get_connection()
+        node = Node(name, role, tags, conn, self._network, 0, 8,
+                    self.public_key, self.private_key)
+        node.boot()
+        self.node_add(node)
+
+    def boot_nodes(self, masters: int = 1, workers: int = 2, offset: int = 0):
+        super().boot_nodes(masters, workers, offset)
+        threads = []
+        for c in range(0, masters):
+            if c == 0:
+                tags = ['master', 'first_master']
+            else:
+                tags = ['master']
+            t = threading.Thread(
+                target=self._boot_node, args=(
+                    f"{self.hardware_uuid}-master-{c}",
+                    NodeRole.MASTER, tags))
+            threads.append(t)
+            t.start()
+
+        for c in range(0, workers):
+            t = threading.Thread(
+                target=self._boot_node, args=(
+                    f"{self.hardware_uuid}-worker-{c}",
+                    NodeRole.WORKER, ['worker']))
+            threads.append(t)
+            t.start()
+
+        # wait for all threads to finish
+        for t in threads:
+            t.join()
+
+    def prepare_nodes(self):
```

Comment:
(nit) I wonder if this can go into BaseHardware

Reply:
Done in an extra commit